### PR TITLE
docs: some cleanup and improvements

### PR DESCRIPTION
**Pull request by gmarull**
Summary:
- Current Makefile is just a thin wrapper around the CMake based documentation build. Removing this Makefile makes things more consistent as CMake is used everywhere in Zephyr as of today.
- Sphinx supports parallelized build (-j) option. Setting it to `auto` makes use of all available cores. This option seems to speed up the build on multi-core machines, even though not much when Kconfig is enabled.

Quick comparison on `-j auto` usage (cpu count=12):

`KCONFIG_TURBO_MODE=0`
- `SPHINXOPTS=`:        `ninja -C doc/_build htmldocs  650.51s user 8.50s system 100% cpu 10:55.39 total`
- `SPHINXOPTS=-j auto`: `ninja -C doc/_build htmldocs  2094.90s user 62.44s system 368% cpu 9:46.23 total`

`KCONFIG_TURBO_MODE=1`:
- `SPHINXOPTS=`:        `ninja -C doc/_build htmldocs  303.29s user 4.48s system 101% cpu 5:04.46 total`
- `SPHINXOPTS=-j auto`: `ninja -C doc/_build htmldocs  1060.54s user 21.20s system 555% cpu 3:14.64 total`

### CODE REVIEW

**Comment by gmarull** [comment]
@galak why does compliance error with `Must be run from the top-level dir. of a kernel tree`?

File: .github/workflows/doc-build.yml [hunk 1]
```diff
@@ -55,8 +54,8 @@ jobs:
 
     - name: build-docs
       run: |
-        source zephyr-env.sh
-        make htmldocs
+        cmake -GNinja -Bdoc/_build doc
```
Comment (carlescufi) [inline .github/workflows/doc-build.yml:57]:
```suggestion
        cmake -GNinja -S doc -Bdoc/build
```

File: .github/workflows/doc-publish.yml [hunk 2]
```diff
@@ -78,8 +78,8 @@ jobs:
       env:
         DOC_TAG: ${{ steps.tag.outputs.TYPE }}
       run: |
-        source zephyr-env.sh
-        make DOC_TAG=${DOC_TAG} htmldocs
+        cmake -GNinja -Bdoc/_build -DDOC_TAG=${DOC_TAG} doc
```
Comment (carlescufi) [inline .github/workflows/doc-publish.yml:81]:
```suggestion
        cmake -GNinja -S doc -Bdoc/_build -DDOC_TAG=${DOC_TAG}
```

File: doc/CMakeLists.txt [hunk 3]
```diff
@@ -59,7 +59,7 @@ if(${LATEXMK} STREQUAL LATEXMK-NOTFOUND)
 endif()
 
 if(NOT DEFINED SPHINXOPTS)
-  set(SPHINXOPTS -q)
+  set(SPHINXOPTS -q -j auto)
```
Comment (carlescufi) [inline doc/CMakeLists.txt:62]: 👍🏻

File: .github/workflows/doc-publish.yml (hunk 2 above)
Comment (gmarull) [reply .github/workflows/doc-publish.yml]: fixed

File: .github/workflows/doc-build.yml (hunk 1 above)
Comment (gmarull) [reply .github/workflows/doc-build.yml]: fixed

**Review by carlescufi — APPROVED** [review]
LGTM, glad to see that `Makefile` go and `-j auto` by default.

**Review by jocelyn-li — APPROVED** [review]

**Comment by carlescufi** [comment]
@nashif can you take a look?

**Review by nashif — CHANGES_REQUESTED** [review]
how is having to type:

```
cmake -GNinja -S doc -Bdoc/_build
ninja -C doc/_build htmldocs
```

instead of just 
```
make
```
an improvement?

**Comment by gmarull** [comment]
> how is having to type:
> 
> ```
> cmake -GNinja -S doc -Bdoc/_build
> ninja -C doc/_build htmldocs
> ```
> 
> instead of just
> 
> ```
> make
> ```

in order to rebuild docs one can just do `ninja htmldocs` everytime when inside the build dir.

> 
> an improvement?

the question here is: why do we want to maintain a top-level Makefile to build the documentation?

**Comment by carlescufi** [comment]
> the question here is: why do we want to maintain a top-level Makefile to build the documentation?

I agree, having a `Makefile` at the root just to build the doc does not make sense in my opinion.

**Comment by nashif** [comment]
> in order to rebuild docs one can just do `ninja htmldocs` everytime when inside the build dir. Actually it's not just `make`, it's `source zephyr-env.sh` and `make htmldocs`...

lets not pull `source zephyr-env.sh` into this please, for me this is a given and is always done as the first thing when I go into the repo, building docs right now is really just calling 'make', the default target gets me what I want.

> I agree, having a `Makefile` at the root just to build the doc does not make sense in my opinion.

Then provide a way or a shortcut (script?) to build docs without having to deal with and remember cmake and ninja options like 
```
cmake -GNinja -S doc -Bdoc/_build
ninja -C doc/_build htmldocs
```

**Comment by gmarull** [comment]
I think that a shortcut script isn't ideal, what if I want to run `doxygen` target? What if I just want to skip CMake configuration step?

Note that CMake command allows for some more simplicity, e.g. taking Ninja out of the game (doesn't add value here other than offering better experience for Windows users):

```shell
# from `doc` folder
cmake -B_build .
cd _build
make htmldocs
# edit files
make htmldocs
...
```

**Comment by carlescufi** [comment]
> Then provide a way or a shortcut (script?) to build docs without having to deal with and remember cmake and ninja options like

We rejected this in-tree back when we switched from `make` to `CMake`, asking people to have their own scripts or aliases, if memory serves.

**Comment by gmarull** [comment]
> > EDIT: @nashif for reference, here is what I am referring to: [#5201 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/5201#pullrequestreview-79993045)
> 
> 
> 
> right, then leave the Makefile. I am really not sure what problem are we trying to solve here.

IMO it's about tidiness. A top-level Makefile (a wrapper Makefile to be precise) just for docs is odd. Would it be possible at least to move it down to docs...?

**Comment by carlescufi** [comment]
> > > EDIT: @nashif for reference, here is what I am referring to: [#5201 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/5201#pullrequestreview-79993045)
> > 
> > 
> > right, then leave the Makefile. I am really not sure what problem are we trying to solve here.
> 
> IMO it's about tidiness. A top-level Makefile (a wrapper Makefile to be precise) just for docs is odd. Even more if Make is not actually used as a build system... Would it be possible at least to move it down to docs...?

I do agree with @gmarull here. That `Makefile` used to be the project's main entry point to the build system, but now it's just a wrapper around CMake for the docs? it's kind of a very strange reasoning for a `Makefile` to be there at all to be honest.

**Comment by carlescufi** [comment]
> > > > EDIT: @nashif for reference, here is what I am referring to: [#5201 (review)](https://github.com/zephyrproject-rtos/zephyr/pull/5201#pullrequestreview-79993045)
> > > 
> > > 
> > > right, then leave the Makefile. I am really not sure what problem are we trying to solve here.
> > 
> > 
> > IMO it's about tidiness. A top-level Makefile (a wrapper Makefile to be precise) just for docs is odd. Even more if Make is not actually used as a build system... Would it be possible at least to move it down to docs...?
> 
> I do agree with @gmarull here. That `Makefile` used to be the project's main entry point to the build system, but now it's just a wrapper around CMake for the docs? it's kind of a very strange reasoning for a `Makefile` to be there at all to be honest.

In fact, I would argue that it's misleading to have that `Makefile` at all. Everything in Zephyr is CMake-based, but users looking at the codebase for the first time see a `Makefile` in the root folder, which is a red herring.

**Comment by carlescufi** [comment]
@nashif do you really want to keep this Makefile? I disagree, but we can discuss it.

**Comment by nashif** [comment]
> @nashif do you really want to keep this Makefile? I disagree, but we can discuss it.

if the makefile being in the root is the problem, just move it under doc, Is this acceptable?

**Comment by carlescufi** [comment]
> > @nashif do you really want to keep this Makefile? I disagree, but we can discuss it.
> 
> if the makefile being in the root is the problem, just move it under doc, Is this acceptable?

If you insist... but I really think that keeping a Makefile that just wraps a CMake script makes no sense.

**Comment by carlescufi** [comment]
dev-review 2021.03.25:

- Move the `Makefile` to `doc/`
- Long term, create a `west doc-build` command

**Comment by gmarull** [comment]
@nashif should be ready now

**Review by nashif — APPROVED** [review]

**Review by mbolivar-nordic — CHANGES_REQUESTED** [review]
@nashif I remember trying this a couple of years ago and dbkinder told me this lead to nondeterministic error output being caught by the .known-issues checks.

@gmarull have you tried this with different `-j` levels (not `-j auto`, but `-j 2`, `-j 4`, etc.) to make sure this doesn't happen?

**Comment by mbolivar-nordic** [comment]
lol, too late to the party

**Comment by gmarull** [comment]
@mbolivar-nordic I've only tried `-j auto` (which is `-j 12` on my machine) and I haven't seen any errors so far. Do you mean that known issues filter is masking some real issues?

**Comment by mbolivar-nordic** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/14983#discussion_r270030987 is the thread where dbkinder mentioned problems

**Comment by mbolivar-nordic** [comment]
> @mbolivar-nordic I've only tried `-j auto` (which is `-j 12` on my machine) and I haven't seen any errors so far. Do you mean that known issues filter is masking some real issues?

I'm reading tea leaves here, but basically I think that we are (or at least at some point, we were) relying on a particular serialization of how the .rst files are processed when masking errors in .known-issues.

For example -- and again I want to emphasize I am just making an educated guess here since it was @dbkinder, the former docs maintainer, who pointed out the error -- if we get a `Duplicate C declaration` warning from `foo.rst` using `-j1`, we might get the same warning from `bar.rst` using `-j3` if `bar.rst` happens to get processed first in `-j3` and processed second in `-j1`.

**Pull request merged**